thirdparty/hnswlib/hnswlib/hnswalg.h

```diff
@@ -210,8 +228,53 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
     char* map_;
     size_t map_size_;
 
+    bool sq_enabled_ = false;
+    bool sq_refine_enabled_ = false;
+    float alpha_ = 0.0f;
+
     mutable knowhere::lru_cache<uint64_t, tableint> lru_cache;
 
+    void
+    trainSQuant(const float* train_data, size_t ntrain) {
+        size_t dim = *(size_t*)dist_func_param_;
+        for (size_t i = 0; i < ntrain; ++i) {
+            const float* vec = train_data + i * dim;
+            std::unique_ptr<float[]> vec_norm = nullptr;
+            if (metric_type_ == Metric::COSINE) {
+                vec_norm = knowhere::CopyAndNormalizeVecs(vec, 1, dim);
+                vec = vec_norm.get();
+            }
+            for (size_t j = 0; j < dim; ++j) {
+                alpha_ = std::max(alpha_, std::abs(vec[j]));
+            }
+        }
+    }
+
+    void
+    encodeSQuant(const float* from, int8_t* to) const {
+        size_t dim = *(size_t*)dist_func_param_;
+        std::unique_ptr<float[]> data_norm = nullptr;
+        if (metric_type_ == Metric::COSINE) {
+            data_norm = knowhere::CopyAndNormalizeVecs(from, 1, dim);
+            from = data_norm.get();
+        }
+        for (size_t i = 0; i < dim; ++i) {
+            float x = from[i] / alpha_;
+            if (x > 1.0f) {
```

Comment:
So, this implementation assumes that the data distribution is symmetric around 0, correct?

Reply:
Yes. It's the most common scenario